### ISSUE TITLE
Maybe solved poor performance on mixed js/ts codebase

### ISSUE THREAD

**Issue by richard-ejem** [typescript-eslint/typescript-eslint#4086]
I've found an interesting solution to problems related to https://github.com/typescript-eslint/typescript-eslint/issues/389 - poor performance on mixed js/ts codebases with `parserOptions.project`. The original issue is unfortunately closed, but I find this highly relevant.

We wanted to introduce `no-unnecessary-type-assertion` rule which requires parserServices, but adding `parserOptions.project` bumped lint time from ~2mins to ~60mins so basically unusable. Then I've found an interesting workaround - add this setting along with all rules depending on it to an `overrides` section only for typescript files.

I hope this can help others facing the same issue, or even better understand the problem with poor performance with tsconfig.json and maybe finding even cleaner soluton.

**Repro**

```JSON
  "overrides": [
    {
      "files": ["**/*.{ts,tsx}"],
      "parserOptions": {
        "project": "./tsconfig.json"
      },
      "rules": {
        "@typescript-eslint/no-unnecessary-type-assertion": ["error"],
      }
    }
  ]
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.31.1` |
| `@typescript-eslint/parser`        | `4.31.1` |
| `TypeScript`                       | `4.4.3` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `12.18.3` |

**Comment by bradzacher** [comment]
Given that your lint time is already ~2min - I'd say that the issue is more that your codebase is just big!
Limiting it to just TS files is going to be a "stop gap" for you - if you onboard more of your codebase to TS, then your times will just increase proportionally.

The unfortunate thing is that at some point a company's codebase scales past just doing `yarn lint`, and needs to start evolving tooling which can better handle it.
For example - at FB in our largest repo we do not use the ESLint "binary" - because otherwise it would literally take hours to lint the entire codebase (WITHOUT type information). Instead have built custom integrations using the ESLint API which allow us to lint smaller parts of the codebase or understand files that should be touched by the changes in a PR.

There's also the steps listed here which can help in some cases:
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#my-linting-feels-really-slow

**Comment by richard-ejem** [comment]
@bradzacher thanks for response! I'm sorry I cannot provide our codebase as a showcase as it is private code, but I can at least provide this info:

- there is about 2000 js/tsx files and 1000 ts/tsx files in the codebase. While providing tsconfig.json to whole .eslintrc makes the lint last mentioned 60 minutes, providing it to just ts/tsx files makes almost no difference, total time is still about 2min. This indicates that there is really some issue with mixed js/ts codebases, and also because of that, I don't expect the time to grow proportionally when we migrate more code to TS.
- my solution doesn't limit linting to ts files only. I limited to typescript files only these rules requiring tsconfig reference, and these rules are likely (maybe always) irrelevant for pure .js anyway. So yes, I feel it is a bit hacky solution, but eslint now checks everything I wanted it to check in both js and ts files.

I tried hints from the FAQ, now considering some slow lint rules to remove, and caching helps a lot. But it still seems to me that the tremendous slowdown after adding tsconfig.json to .eslintrc is an issue itself.
